From 4d3f98d83d4cd464c220bbfd290d44e5e2cb5227 Mon Sep 17 00:00:00 2001
From: Bharath Ramesh <rameshb@vmware.com>
Date: Tue, 21 Oct 2014 17:03:36 -0700
Subject: [PATCH 4/7] Fix shell extension to build using newer Windows SDK

Microsoft decided to rename IAsyncOperation to IDataObjectAsyncCapability in
their Windows SDK.

NOTE: These changes are untested

Reviewed by: akreher, arolett
Review URL: https://reviewboard.eng.vmware.com/r/686986/
---
 com/win32com/src/include/PythonCOMRegister.h    |  1 +
 com/win32comext/shell/src/PyIAsyncOperation.cpp | 24 ++++++++++++------------
 com/win32comext/shell/src/PyIAsyncOperation.h   |  6 +++---
 com/win32comext/shell/src/shell.cpp             |  4 +++-
 4 files changed, 19 insertions(+), 16 deletions(-)

diff --git a/com/win32com/src/include/PythonCOMRegister.h b/com/win32com/src/include/PythonCOMRegister.h
index 1cc3b14..95fe8b3 100644
--- a/com/win32com/src/include/PythonCOMRegister.h
+++ b/com/win32com/src/include/PythonCOMRegister.h
@@ -27,6 +27,7 @@ typedef struct
 #define PYCOM_INTERFACE_CLIENT_ONLY(ifc)	{ &IID_I##ifc, "I" #ifc, "IID_I" #ifc, &PyI##ifc::type, NULL }
 #define PYCOM_INTERFACE_SERVER_ONLY(ifc)	{ &IID_I##ifc, "I" #ifc, "IID_I" #ifc, NULL, GET_PYGATEWAY_CTOR(PyG##ifc)}
 #define PYCOM_INTERFACE_FULL(ifc)	{ &IID_I##ifc, "I" #ifc, "IID_I" #ifc, &PyI##ifc::type, GET_PYGATEWAY_CTOR(PyG##ifc)}
+#define PYCOM_INTERFACE_FULL_ASYNC(sdk_ifc,ifc)	{ &IID_I##sdk_ifc, "I" #ifc, "IID_I" #ifc, &PyI##ifc::type, GET_PYGATEWAY_CTOR(PyG##ifc)}
 
 // Versions that use __uuidof() to get the IID, which seems to avoid the need
 // to link with a lib holding the IIDs.  Note that almost all extensions
diff --git a/com/win32comext/shell/src/PyIAsyncOperation.cpp b/com/win32comext/shell/src/PyIAsyncOperation.cpp
index 8c53b60..a8ed93a 100644
--- a/com/win32comext/shell/src/PyIAsyncOperation.cpp
+++ b/com/win32comext/shell/src/PyIAsyncOperation.cpp
@@ -19,15 +19,15 @@ PyIAsyncOperation::~PyIAsyncOperation()
 {
 }
 
-/* static */ IAsyncOperation *PyIAsyncOperation::GetI(PyObject *self)
+/* static */ IDataObjectAsyncCapability *PyIAsyncOperation::GetI(PyObject *self)
 {
-	return (IAsyncOperation *)PyIUnknown::GetI(self);
+	return (IDataObjectAsyncCapability *)PyIUnknown::GetI(self);
 }
 
 // @pymethod |PyIAsyncOperation|SetAsyncMode|Description of SetAsyncMode.
 PyObject *PyIAsyncOperation::SetAsyncMode(PyObject *self, PyObject *args)
 {
-	IAsyncOperation *pIAO = GetI(self);
+	IDataObjectAsyncCapability *pIAO = GetI(self);
 	if ( pIAO == NULL )
 		return NULL;
 	// @pyparm int|fDoOpAsync||Description for fDoOpAsync
@@ -41,7 +41,7 @@ PyObject *PyIAsyncOperation::SetAsyncMode(PyObject *self, PyObject *args)
 	PY_INTERFACE_POSTCALL;
 
 	if ( FAILED(hr) )
-		return PyCom_BuildPyException(hr, pIAO, IID_IAsyncOperation );
+		return PyCom_BuildPyException(hr, pIAO, IID_IDataObjectAsyncCapability );
 	Py_INCREF(Py_None);
 	return Py_None;
 
@@ -50,7 +50,7 @@ PyObject *PyIAsyncOperation::SetAsyncMode(PyObject *self, PyObject *args)
 // @pymethod bool|PyIAsyncOperation|GetAsyncMode|Description of GetAsyncMode.
 PyObject *PyIAsyncOperation::GetAsyncMode(PyObject *self, PyObject *args)
 {
-	IAsyncOperation *pIAO = GetI(self);
+	IDataObjectAsyncCapability *pIAO = GetI(self);
 	if ( pIAO == NULL )
 		return NULL;
 	BOOL fIsOpAsync;
@@ -61,7 +61,7 @@ PyObject *PyIAsyncOperation::GetAsyncMode(PyObject *self, PyObject *args)
 	hr = pIAO->GetAsyncMode( &fIsOpAsync );
 	PY_INTERFACE_POSTCALL;
 	if ( FAILED(hr) )
-		return PyCom_BuildPyException(hr, pIAO, IID_IAsyncOperation );
+		return PyCom_BuildPyException(hr, pIAO, IID_IDataObjectAsyncCapability );
 	PyObject *rc = fIsOpAsync ? Py_True : Py_False;
 	return rc;
 }
@@ -69,7 +69,7 @@ PyObject *PyIAsyncOperation::GetAsyncMode(PyObject *self, PyObject *args)
 // @pymethod |PyIAsyncOperation|StartOperation|Description of StartOperation.
 PyObject *PyIAsyncOperation::StartOperation(PyObject *self, PyObject *args)
 {
-	IAsyncOperation *pIAO = GetI(self);
+	IDataObjectAsyncCapability *pIAO = GetI(self);
 	if ( pIAO == NULL )
 		return NULL;
 	// @pyparm <o PyIBindCtx>|pbcReserved||Description for pbcReserved
@@ -89,7 +89,7 @@ PyObject *PyIAsyncOperation::StartOperation(PyObject *self, PyObject *args)
 	PY_INTERFACE_POSTCALL;
 
 	if ( FAILED(hr) )
-		return PyCom_BuildPyException(hr, pIAO, IID_IAsyncOperation );
+		return PyCom_BuildPyException(hr, pIAO, IID_IDataObjectAsyncCapability );
 	Py_INCREF(Py_None);
 	return Py_None;
 
@@ -98,7 +98,7 @@ PyObject *PyIAsyncOperation::StartOperation(PyObject *self, PyObject *args)
 // @pymethod |PyIAsyncOperation|InOperation|Description of InOperation.
 PyObject *PyIAsyncOperation::InOperation(PyObject *self, PyObject *args)
 {
-	IAsyncOperation *pIAO = GetI(self);
+	IDataObjectAsyncCapability *pIAO = GetI(self);
 	if ( pIAO == NULL )
 		return NULL;
 	BOOL fInAsyncOp;
@@ -109,7 +109,7 @@ PyObject *PyIAsyncOperation::InOperation(PyObject *self, PyObject *args)
 	hr = pIAO->InOperation( &fInAsyncOp );
 	PY_INTERFACE_POSTCALL;
 	if ( FAILED(hr) )
-		return PyCom_BuildPyException(hr, pIAO, IID_IAsyncOperation );
+		return PyCom_BuildPyException(hr, pIAO, IID_IDataObjectAsyncCapability );
 	PyObject *rc = fInAsyncOp ? Py_True : Py_False;
 	return rc;
 }
@@ -117,7 +117,7 @@ PyObject *PyIAsyncOperation::InOperation(PyObject *self, PyObject *args)
 // @pymethod |PyIAsyncOperation|EndOperation|Description of EndOperation.
 PyObject *PyIAsyncOperation::EndOperation(PyObject *self, PyObject *args)
 {
-	IAsyncOperation *pIAO = GetI(self);
+	IDataObjectAsyncCapability *pIAO = GetI(self);
 	if ( pIAO == NULL )
 		return NULL;
 	// @pyparm int|hResult||Description for hResult
@@ -141,7 +141,7 @@ PyObject *PyIAsyncOperation::EndOperation(PyObject *self, PyObject *args)
 	PY_INTERFACE_POSTCALL;
 
 	if ( FAILED(hr) )
-		return PyCom_BuildPyException(hr, pIAO, IID_IAsyncOperation );
+		return PyCom_BuildPyException(hr, pIAO, IID_IDataObjectAsyncCapability );
 	Py_INCREF(Py_None);
 	return Py_None;
 
diff --git a/com/win32comext/shell/src/PyIAsyncOperation.h b/com/win32comext/shell/src/PyIAsyncOperation.h
index 1122609..5df67c5 100644
--- a/com/win32comext/shell/src/PyIAsyncOperation.h
+++ b/com/win32comext/shell/src/PyIAsyncOperation.h
@@ -8,7 +8,7 @@ class PyIAsyncOperation : public PyIUnknown
 {
 public:
 	MAKE_PYCOM_CTOR(PyIAsyncOperation);
-	static IAsyncOperation *GetI(PyObject *self);
+	static IDataObjectAsyncCapability *GetI(PyObject *self);
 	static PyComTypeObject type;
 
 	// The Python methods
@@ -26,11 +26,11 @@ protected:
 //
 // Gateway Declaration
 
-class PyGAsyncOperation : public PyGatewayBase, public IAsyncOperation
+class PyGAsyncOperation : public PyGatewayBase, public IDataObjectAsyncCapability
 {
 protected:
 	PyGAsyncOperation(PyObject *instance) : PyGatewayBase(instance) { ; }
-	PYGATEWAY_MAKE_SUPPORT2(PyGAsyncOperation, IAsyncOperation, IID_IAsyncOperation, PyGatewayBase)
+	PYGATEWAY_MAKE_SUPPORT2(PyGAsyncOperation, IDataObjectAsyncCapability, IID_IDataObjectAsyncCapability, PyGatewayBase)
 
 
 
diff --git a/com/win32comext/shell/src/shell.cpp b/com/win32comext/shell/src/shell.cpp
index 8b6e78c..889353e 100644
--- a/com/win32comext/shell/src/shell.cpp
+++ b/com/win32comext/shell/src/shell.cpp
@@ -3580,7 +3580,9 @@ static const PyCom_InterfaceSupportInfo g_interfaceSupportData[] =
 	PYCOM_INTERFACE_IID_ONLY		  (ShellLinkA),
 	PYCOM_INTERFACE_CLSID_ONLY		  (ShellLink),
 	PYCOM_INTERFACE_IID_ONLY		  (ShellLinkW),
-	PYCOM_INTERFACE_FULL(AsyncOperation),
+	// rameshb: MS renamed AsyncOperation to DataObjectAsyncCapability
+	// This allows AsyncOperation interface to call DataObjectAsyncCapability
+	PYCOM_INTERFACE_FULL_ASYNC(DataObjectAsyncCapability, AsyncOperation),
 	PYCOM_INTERFACE_FULL(ContextMenu),
 	PYCOM_INTERFACE_SERVER_ONLY(ContextMenu2),
 	PYCOM_INTERFACE_SERVER_ONLY(ContextMenu3),
-- 
2.8.0.windows.1

